File: assets/design_handoff_home_room/assets/bit-pad/bitpad-light.js
/* ==========================================================================
   BIT HOVER PAD — pixel-art light engine.
   Renders a chunky, ordered-dithered cyan glow on a LOW-RES canvas that CSS
   then scales up nearest-neighbour (image-rendering:pixelated) — so the light
   is authentic pixel art, never a smooth CSS gradient.

   The field is a radial focus at the centre that spreads out widely (set rx
   wide so it reads wider than BIT) and pools downward on the floor (ryUp keeps
   the halo from climbing the pillars). Animated with a slow breathing fade + a
   flicker that drops chunks of pixels out (authentic, not just opacity).
   Reduce-motion safe.

   NOTE: the pad→BIT tether beam is intentionally NOT part of this engine — it
   is designed as a separate element.

   Usage:
     <canvas id="padGlow"></canvas>   // sized + positioned in CSS
     BitPadLight.init(canvas, {
       cols, rows,          // low-res cell grid (canvas.width/height)
       cx, cy, rx, ry,      // focus centre + radii, in cells (ry = downward)
       ryUp,                // optional tighter upward radius (default = ry)
       fps                  // default 14 (chunky)
     });
   ========================================================================== */
(function () {
  const BAYER = [
    [0, 8, 2, 10],
    [12, 4, 14, 6],
    [3, 11, 1, 9],
    [15, 7, 13, 5],
  ];
  // cyan ramp — deep → bright → near-white core (rgb, base alpha)
  const TIERS = [
    null,
    { c: '30,130,178', a: 0.24 },
    { c: '48,190,232', a: 0.46 },
    { c: '132,232,255', a: 0.70 },
    { c: '214,250,255', a: 0.88 },
  ];

  function init(canvas, cfg) {
    const ctx = canvas.getContext('2d');
    const COLS = cfg.cols, ROWS = cfg.rows;
    canvas.width = COLS; canvas.height = ROWS;
    ctx.imageSmoothingEnabled = false;
    const cx = cfg.cx, cy = cfg.cy, rx = cfg.rx, ry = cfg.ry;
    const ryUp = cfg.ryUp || ry;          // tighter upward = floor pool, not a halo up the pillars
    const frameMs = 1000 / (cfg.fps || 14);

    function field(x, y) {
      // radial focus — wide horizontally, asymmetric vertically (pools downward)
      const dx = (x - cx) / rx;
      const dy = (y < cy) ? (y - cy) / ryUp : (y - cy) / ry;
      let v = 1 - Math.hypot(dx, dy);
      if (v < 0.10) v = 0;                 // trim stray far specks
      return v < 0 ? 0 : v;
    }

    function draw(intensity) {
      ctx.clearRect(0, 0, COLS, ROWS);
      for (let y = 0; y < ROWS; y++) {
        for (let x = 0; x < COLS; x++) {
          let v = field(x, y) * intensity;
          if (v <= 0) continue;
          const th = (BAYER[y & 3][x & 3] + 0.5) / 16;
          let q = v * 4, lvl = Math.floor(q);
          if (q - lvl > th) lvl++;
          if (lvl <= 0) continue;
          if (lvl > 4) lvl = 4;
          const t = TIERS[lvl];
          ctx.fillStyle = 'rgba(' + t.c + ',' + t.a + ')';
          ctx.fillRect(x, y, 1, 1);
        }
      }
    }

    const reduce = window.matchMedia &&
      window.matchMedia('(prefers-reduced-motion: reduce)').matches;
    if (reduce) { draw(0.85); return; }

    let flick = 0.85, t = 0, dipUntil = 0, nextDipRoll = 0, last = 0;
    function loop(now) {
      requestAnimationFrame(loop);
      if (now - last < frameMs) return;
      const dt = (now - last) / 1000; last = now; t += dt;
      // slow breathing fade
      const breath = 0.80 + 0.20 * Math.sin(t * 1.6);
      // flicker: per-frame jitter + occasional brief dropouts
      let target = breath * (0.94 + Math.random() * 0.06);
      if (now > nextDipRoll) { nextDipRoll = now + 90; if (Math.random() < 0.16) dipUntil = now + 50 + Math.random() * 130; }
      if (now < dipUntil) target *= 0.5 + Math.random() * 0.22;
      flick += (target - flick) * 0.55;
      draw(flick);
    }
    requestAnimationFrame(loop);
  }

  window.BitPadLight = { init };
})();

File: lib/widgets/room/bit_pad_light.dart
import 'dart:math' as math;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

/// Pixel-art floor-pool light under BIT's emitter pad — a verbatim port of the
/// prototype's `bitpad-light.js`. A chunky **Bayer-4×4 ordered-dithered** cyan
/// radial rendered on a low-res cell grid and scaled up nearest-neighbour, so
/// the light reads as authentic pixel art, never a smooth gradient
/// (handoff GUARDRAILS #1).
///
/// The field pools **wide** horizontally (`_rx`) and **downward** on the floor
/// (`_ryUp` tighter than `_ry` keeps the halo from climbing the pad's pillars).
/// Slow breathing fade + occasional pixel-dropout flicker, quantised to ~14fps
/// (chunky on purpose). Reduced motion → a static lit frame at 0.85.
///
/// The cyan ramp below is canonical procedural-art (a light ramp), not brand
/// color — same status as `bit_boot.dart`'s `_metal` map.
class BitPadLightPainter extends CustomPainter {
  BitPadLightPainter({
    required this.time,
    required this.reduceMotion,
    this.tint = 0.0,
  }) : super(repaint: time);

  /// Shared room clock, in seconds.
  final ValueListenable<double> time;
  final bool reduceMotion;

  /// Underglow hue, 0→1: 0 = BIT's turquoise (home), 1 = magenta (a haul on the
  /// pad). The homecoming tints to magenta on deposit; collect reverses it. Each
  /// dither tier lerps turquoise→magenta keeping its tier alpha.
  final double tint;

  // Config — verbatim from preview.html's BitPadLight.init.
  static const int _cols = 68, _rows = 60;
  static const double _cx = 34, _cy = 40, _rx = 30, _ry = 22, _ryUp = 10;

  static const List<List<int>> _bayer = [
    [0, 8, 2, 10],
    [12, 4, 14, 6],
    [3, 11, 1, 9],
    [15, 7, 13, 5],
  ];

  // BIT turquoise ramp: deep → bright → near-white core (color carries the tier
  // alpha). Recoloured off recovery-cyan so BIT's machine light owns its own hue.
  static const List<Color?> _tiers = [
    null,
    Color(0x3D1A968E),
    Color(0x7528CEC2),
    Color(0xB380F0E4),
    Color(0xE0D2FFFA),
  ];

  // Magenta haul ramp (sampled from gem.png: deep facet → mid → core →
  // highlight) at the same per-tier alphas, so a haul recolours the same dither
  // without changing its shape. Canonical procedural-art palette, not brand.
  static const List<Color?> _magentaTiers = [
    null,
    Color(0x3D961C8C),
    Color(0x75E028A0),
    Color(0xB3FF4DCD),
    Color(0xE0FF96E6),
  ];

  double _field(int x, int y) {
    final dx = (x - _cx) / _rx;
    final dy = (y < _cy) ? (y - _cy) / _ryUp : (y - _cy) / _ry;
    var v = 1 - math.sqrt(dx * dx + dy * dy);
    if (v < 0.10) v = 0; // trim stray far specks
    return v < 0 ? 0 : v;
  }

  // Deterministic 0..1 hash noise (golden-stable; no per-frame RNG state).
  double _noise(double t) {
    final s = math.sin(t * 12.9898) * 43758.5453;
    return s - s.floorToDouble();
  }

  double _intensity() {
    if (reduceMotion) return 0.85;
    final t = time.value;
    final tq = (t * 14).floorToDouble() / 14; // chunky ~14fps
    final breath = 0.80 + 0.20 * math.sin(tq * 1.6);
    final jitter = 0.94 + 0.06 * _noise(tq * 1.7);
    final dip = _noise(tq * 0.37 + 11) < 0.16
        ? (0.5 + 0.22 * _noise(tq * 2.1 + 3))
        : 1.0;
    return breath * jitter * dip;
  }

  @override
  void paint(Canvas canvas, Size size) {
    final cw = size.width / _cols, ch = size.height / _rows;
    final intensity = _intensity();
    final paint = Paint()..isAntiAlias = false;
    for (var y = 0; y < _rows; y++) {
      for (var x = 0; x < _cols; x++) {
        final v = _field(x, y) * intensity;
        if (v <= 0) continue;
        final th = (_bayer[y & 3][x & 3] + 0.5) / 16;
        final q = v * 4;
        var lvl = q.floor();
        if (q - lvl > th) lvl++;
        if (lvl <= 0) continue;
        if (lvl > 4) lvl = 4;
        var c = _tiers[lvl];
        if (c == null) continue;
        if (tint > 0) c = Color.lerp(c, _magentaTiers[lvl], tint)!;
        paint.color = c;
        canvas.drawRect(Rect.fromLTWH(x * cw, y * ch, cw, ch), paint);
      }
    }
  }

  @override
  bool shouldRepaint(covariant BitPadLightPainter old) =>
      old.reduceMotion != reduceMotion || old.tint != tint;
}
